Home Work solution js-salary-info

Danila2006 opened the pull request:
Created new file DataGettingException.java; Modified SalaryInfo.java

mateacademy-ai-mentor reviewed (commented):
Sorry, the pull request review was unsuccessful for the following reasons:

> [!WARNING]
> **The test task has already been approved.**
> <sub>*[CODE: 9]*</sub>


If the issue is within your control, please address it.
The next review attempt <ins>will automatically launch</ins> in 1 hour.
If you believe this is an error or need assistance, please contact the support team.

mateacademy-ai-mentor reviewed (commented):
Sorry, the pull request review was unsuccessful for the following reasons:

> [!WARNING]
> **The test task has already been approved.**
> <sub>*[CODE: 9]*</sub>


If the issue is within your control, please address it.
The next review attempt <ins>will automatically launch</ins> in 1 hour.
If you believe this is an error or need assistance, please contact the support team.

mateacademy-ai-mentor reviewed (commented):
Sorry, the pull request review was unsuccessful for the following reasons:

> [!WARNING]
> **The test task has already been approved.**
> <sub>*[CODE: 9]*</sub>


If the issue is within your control, please address it.
The next review attempt <ins>will automatically launch</ins> in 1 hour.
If you believe this is an error or need assistance, please contact the support team.

mateacademy-ai-mentor reviewed (commented):
Sorry, the pull request review was unsuccessful for the following reasons:

> [!WARNING]
> **The test task has already been approved.**
> <sub>*[CODE: 9]*</sub>


If the issue is within your control, please address it.
The next review attempt <ins>will automatically launch</ins> in 1 hour.
If you believe this is an error or need assistance, please contact the support team.

Elena-Bruyako reviewed (changes requested):
fix build first

Danila2006 commented:
Can I change the tests, for example, flag an exception that might be returned by a method

Elena-Bruyako reviewed (changes requested):
fix build

Danila2006 commented:
```java
package core.basesyntax;

import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.Arrays;

public class SalaryInfo {
    public String getSalaryInfo(String[] names, String[] data, String dateFrom, String dateTo) {

        if (names == null || data == null || dateFrom == null
                || dateTo == null) {
            throw new DataGettingException("One or more input arguments are null");
        }

        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("dd.MM.yyyy");

        LocalDate newDateFrom;
        LocalDate newDateTo;

        try {
            newDateFrom = LocalDate.parse(dateFrom, formatter);
            newDateTo = LocalDate.parse(dateTo, formatter);
        } catch (Exception e) {
            throw new DataGettingException("Invalid date format. Excepted format: dd.MM.yyyy");
        }

        if (newDateFrom.getYear() > newDateTo.getYear()
                || (newDateFrom.getYear() == newDateTo.getYear()
                && newDateFrom.getMonthValue() > newDateTo.getMonthValue())
                || (newDateFrom.getYear() == newDateTo.getYear()
                && newDateFrom.getMonthValue() == newDateTo.getMonthValue()
                && newDateFrom.getDayOfMonth() > newDateTo.getDayOfMonth())) {
            throw new DataGettingException("Start date cannot be later than end date");
        }

        if (newDateFrom.equals(newDateTo)) {
            return returnInfo(ifEqual(names), dateFrom, dateTo);
        }

        String[] startResult = new String[data.length];

        for (int i = 0; i < data.length; i++) {
            String[] dataArrayI = data[i].split(" ");

            LocalDate date;

            try {
                date = LocalDate.parse(dataArrayI[0], formatter);
            } catch (Exception e) {
                throw new DataGettingException("Invalid date in data entry: " + data[i]);
            }

            boolean withinDateRange = date.isAfter(newDateFrom) && date.isBefore(newDateTo)
                    || date.isEqual(newDateFrom) && date.isBefore(newDateTo)
                    || date.isAfter(newDateFrom) && date.isEqual(newDateTo);

            if (withinDateRange) {
                String name = dataArrayI[1];
                int hours = Integer.parseInt(dataArrayI[2]);
                int money = Integer.parseInt(dataArrayI[3]);
                int salary = hours * money;

                addPersonToStartResult(startResult, name, salary);
            }
        }

        return returnInfo(makeInOrder(startResult, names), dateFrom, dateTo);
    }

    public String returnInfo(String[] array, String dataF, String dataT) {
        StringBuilder str = new StringBuilder();
        str.append("Report for period ").append(dataF).append(" - ").append(dataT).append("\n");

        for (String element : array) {
            str.append(element).append("\n");
        }

        return str.toString().trim();
    }

    public void addPersonToStartResult(String[] array, String name, int salary) {
        boolean personExist = false;

        if (array[0] != null) {
            for (int i = 0; i < array.length; i++) {
                if (array[i] != null) {
                    String[] parts = array[i].split(" - ");
                    if (parts[0].equals(name)) {
                        int currentSalary = Integer.parseInt(parts[1]);
                        currentSalary += salary;
                        parts[1] = String.valueOf(currentSalary);
                        array[i] = String.join(" - ", parts);
                        personExist = true;
                        break;
                    }
                } else {
                    break;
                }
            }
        }

        if (!personExist) {
            for (int i = 0; i < array.length; i++) {
                if (array[i] == null) {
                    array[i] = name + " - " + salary;
                    break;
                }
            }
        }
    }

    public String[] makeInOrder(String[] arrayToOrder, String[] names) {
        String[] orderedArray = new String[names.length];
        String[] beforeSorting = new String[names.length];
        int num = 0;

        for (int g = 0; g < arrayToOrder.length; g++) {
            if (arrayToOrder[g] != null) {
                beforeSorting[num] = arrayToOrder[g];
                num++;
            }
        }

        beforeSorting = Arrays.copyOf(beforeSorting, num);

        for (int i = 0; i < names.length; i++) {
            String currentName = names[i];

            for (int j = 0; j < beforeSorting.length; j++) {
                String[] afterSplit = beforeSorting[j].split(" - ");
                String searchName = afterSplit[0];

                if (currentName.equals(searchName)) {
                    orderedArray[i] = beforeSorting[j];
                    break;
                }
            }
        }

        return orderedArray;
    }

    public String[] ifEqual(String[] names) {
        String[] resultArray = new String[names.length];
        for (int i = 0; i < names.length; i++) {
            resultArray[i] = names[i] + " - 0";
        }

        return resultArray;
    }
}
```

PS D:\Project_Mate_Academy\jv-salary-info> mvn clean package
[INFO] Scanning for projects...
[INFO]
[INFO] ----------------------< mate-academy:salaryInfo >-----------------------
[INFO] Building salaryInfo 1.0-SNAPSHOT
[INFO]   from pom.xml
[INFO] --------------------------------[ jar ]---------------------------------
[INFO]
[INFO] --- clean:3.2.0:clean (default-clean) @ salaryInfo ---
[INFO] Deleting D:\Project_Mate_Academy\jv-salary-info\target
[INFO]
[INFO] --- resources:3.3.1:resources (default-resources) @ salaryInfo ---
[INFO] skip non existing resourceDirectory D:\Project_Mate_Academy\jv-salary-info\src\main\resources
[INFO]
[INFO] --- compiler:3.8.0:compile (default-compile) @ salaryInfo ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 3 source files to D:\Project_Mate_Academy\jv-salary-info\target\classes
[INFO] 
[INFO] --- checkstyle:3.3.0:check (default) @ salaryInfo ---
[INFO] Starting audit...
Audit done.
[INFO] You have 0 Checkstyle violations.
[INFO] 
[INFO] --- resources:3.3.1:testResources (default-testResources) @ salaryInfo ---
[INFO] skip non existing resourceDirectory D:\Project_Mate_Academy\jv-salary-info\src\test\resources
[INFO] 
[INFO] --- compiler:3.8.0:testCompile (default-testCompile) @ salaryInfo ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 1 source file to D:\Project_Mate_Academy\jv-salary-info\target\test-classes
[INFO] 
[INFO] --- surefire:3.2.5:test (default-test) @ salaryInfo ---
[INFO] Using auto detected provider org.apache.maven.surefire.junit4.JUnit4Provider
[INFO] 
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running core.basesyntax.SalaryInfoTest
[ERROR] Tests run: 2, Failures: 2, Errors: 0, Skipped: 0, Time elapsed: 0.109 s <<< FAILURE! -- in core.basesyntax.SalaryInfoTest
[ERROR] core.basesyntax.SalaryInfoTest.getSalaryInfoByTwoMonths -- Time elapsed: 0.032 s <<< FAILURE!
org.junit.ComparisonFailure:
Test failed from date 14.07.2019 to 10.08.2019
actual:
Report for period 14.07.2019 - 10.08.2019
John - 700
Andrew - 1200
Kate - 2140
expected:
Report for period 14.07.2019 - 10.08.2019
John - 700
Andrew - 1200
Kate - 2140 expected:<...07.2019 - 10.08.2019[
John - 700
]ndrew - 1200
Kate - 2140> but was:<...07.2019 - 10.08.2019[
John - 700
Andrew - 1200]
Kate - 2140>
        at org.junit.Assert.assertEquals(Assert.java:115)
        at core.basesyntax.SalaryInfoTest.getSalaryInfoByTwoMonths(SalaryInfoTest.java:105)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
        at java.base/java.lang.reflect.Method.invoke(Method.java:580)
        at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
        at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
        at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
        at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
        at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
        at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
        at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
        at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
        at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
        at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
        at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
        at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:316)
        at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:240)
        at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:214)
        at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:155)
        at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:385)
        at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:162)
        at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:507)
        at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:495)

[ERROR] core.basesyntax.SalaryInfoTest.getSalaryInfoByMonth -- Time elapsed: 0 s <<< FAILURE!
org.junit.ComparisonFailure:
Test failed from date 24.04.2019 to 24.04.2019
actual:
Report for period 24.04.2019 - 24.04.2019
John - 0
Andrew - 0
Kate - 0
expected:
Report for period 24.04.2019 - 24.04.2019
John - 0
Andrew - 0
Kate - 0 expected:<...04.2019 - 24.04.2019[
John - 0
]ndrew - 0
Kate - 0> but was:<...04.2019 - 24.04.2019[
John - 0
Andrew - 0]
Kate - 0>
        at org.junit.Assert.assertEquals(Assert.java:115)
        at core.basesyntax.SalaryInfoTest.getSalaryInfoByMonth(SalaryInfoTest.java:89)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
        at java.base/java.lang.reflect.Method.invoke(Method.java:580)
        at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
        at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
        at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
        at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
        at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
        at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
        at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
        at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
        at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
        at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
        at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
        at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:316)
        at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:240)
        at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:214)
        at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:155)
        at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:385)
        at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:162)
        at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:507)
        at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:495)

[INFO] 
[INFO] Results:
[INFO]
[ERROR] Failures: 
[ERROR]   SalaryInfoTest.getSalaryInfoByMonth:89 Test failed from date 24.04.2019 to 24.04.2019
actual: 
Report for period 24.04.2019 - 24.04.2019
John - 0
Andrew - 0
Kate - 0
expected: 
Report for period 24.04.2019 - 24.04.2019
John - 0
Andrew - 0
Kate - 0 expected:<...04.2019 - 24.04.2019[
John - 0
]ndrew - 0
Kate - 0> but was:<...04.2019 - 24.04.2019[
John - 0
Andrew - 0]
Kate - 0>
[ERROR]   SalaryInfoTest.getSalaryInfoByTwoMonths:105 Test failed from date 14.07.2019 to 10.08.2019
actual: 
Report for period 14.07.2019 - 10.08.2019
John - 700
Andrew - 1200
Kate - 2140
expected: 
Report for period 14.07.2019 - 10.08.2019
John - 700
Andrew - 1200
Kate - 2140 expected:<...07.2019 - 10.08.2019[
John - 700
]ndrew - 1200
Kate - 2140> but was:<...07.2019 - 10.08.2019[
John - 700
Andrew - 1200]
Kate - 2140>
[INFO]
[ERROR] Tests run: 2, Failures: 2, Errors: 0, Skipped: 0
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  6.090 s
[INFO] Finished at: 2024-11-19T18:40:01+01:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:3.2.5:test (default-test) on project salaryInfo: There are test failures.
[ERROR]
[ERROR] Please refer to D:\Project_Mate_Academy\jv-salary-info\target\surefire-reports for the individual test results.
[ERROR] Please refer to dump files (if any exist) [date].dump, [date]-jvmRun[N].dump and [date].dumpstream.
[ERROR] -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
PS D:\Project_Mate_Academy\jv-salary-info>


I can't figure out where the error is, can you please help me?

liliia-ponomarenko reviewed (changes requested):
Unfortunately, your task doesn't pass the basic tests. Let's solve it in the platform chat. Describe your problem with screenshots in our chat.

okuzan commented on src/main/java/core/basesyntax/SalaryInfo.java:
```diff
@@ -1,7 +1,154 @@
+            throw new DataGettingException("Invalid date format. Excepted format: dd.MM.yyyy");
+        }
+
+        if (newDateFrom.getYear() > newDateTo.getYear()
+                || (newDateFrom.getYear() == newDateTo.getYear()
+                && newDateFrom.getMonthValue() > newDateTo.getMonthValue())
+                || (newDateFrom.getYear() == newDateTo.getYear()
+                && newDateFrom.getMonthValue() == newDateTo.getMonthValue()
+                && newDateFrom.getDayOfMonth() > newDateTo.getDayOfMonth())) {
```
use LocalDate's isBefore, isAfter

okuzan reviewed (changes requested):
your solution can be vastly simplified
use only two for loops, extract single method for comparing dates. Extract DateTimeFormatter to a class constant - checklist issue
Utility methods should be private
